Release Automation [Milestone 7]: Automation for initial release preparation

ankita240796 opened the pull request:
## Explanation
- Adds automation for initial release preparation.
- Adds test for `update_indexes.py`, `cut_release_or_hotfix_branch.py`.
- Updates codeowners.
- Adds more instructions for changelog updates.
- Fixes #8027 
- Adds automation to detect schema changes and find jobs to run based on that

## Checklist
- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python -m scripts.pre_commit_linter` and `python -m scripts.run_frontend_tests`.
- [x] The PR is made from a branch that's **not** called "develop".
- [ ] The PR has an appropriate "PROJECT: ..." label (Please add this label for the first-pass review of the PR).
- [x] The PR has an appropriate "CHANGELOG: ..." label (If you are unsure of which label to add, ask the reviewers for guidance).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR addresses the points mentioned in the codeowner checks for the files/folders changed. (See the [codeowner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).)
- [x] The PR is **assigned** to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer and don't tick this checkbox.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue. Do not only request the review but also add the reviewer as an assignee.

ankita240796 committed:
Add automation for release docs

ankita240796 committed:
Update codeowners

ankita240796 committed:
Add tests for cut release or hotfix branch

ankita240796 committed:
Add credentials to gitignore

ankita240796 committed:
Add tests for update indexes

ankita240796 committed:
Add tests for initial release prep

ankita240796 committed:
Make update simpler

ankita240796 committed:
Fix import error

kevinlee12 commented on .github/CODEOWNERS:
```diff
@@ -388,6 +388,31 @@
 /data/ @nithusha21
 
 
+# Release team.
```
It looks like I'm also jointly owning this section as well, can you expand out https://github.com/oppia/oppia/blob/1158db9d40eeee9d50cc1d21a14c4ba9f05e8494/.github/CODEOWNERS#L102 this section?

ankita240796 replied:
Done.

seanlip commented:
Hi @ankita240796 -- please note, automated tests are failing.

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into doc-automate

ankita240796 committed:
Fix tests

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into doc-automate

ankita240796 committed:
Call release info from update script

ankita240796 committed:
Fix name

ankita240796 committed:
Update backup restoration

ankita240796 committed:
Fix test

ankita240796 committed:
Add function to cancel restoration

vojtechjelinek commented on scripts/deploy.py:
```diff
@@ -319,9 +319,6 @@ def check_breakage(app_name, current_release_version):
-    release_journal_url = (
-        'https://drive.google.com/drive/folders/'
-        '0B9KSjiibL_WDNjJyYlEtbTNvY3c')
     issue_filing_url = 'https://github.com/oppia/oppia/milestone/39'
```
Should these also be moved to `release_constants`?

ankita240796 replied:
Yes, done.

vojtechjelinek commented on scripts/initial_release_prep.py:
```diff
@@ -0,0 +1,223 @@
+            'release before starting with the release process.')
+    try:
+        f = python_utils.open_file(RELEASE_CREDENTIALS_FILEPATH, 'w')
+        f.close()
```
Could you add a comment here that this is done just to create an empty file? It is not really intuitive and I initially thought that this are just some forgotten lines of code.

ankita240796 replied:
Done.

vojtechjelinek commented on scripts/initial_release_prep.py:
```diff
@@ -0,0 +1,223 @@
+                RELEASE_CREDENTIALS_FILEPATH))
+        client = pygsheets.authorize(client_secret=RELEASE_CREDENTIALS_FILEPATH)
+
+        existing_jobs_sheet = client.open(
```
Is this sheet even used anymore? I think that the new one is used for 3 last months.

ankita240796 replied:
Yes, it is used for jobs which are already present in codebase and not newly added like Migration jobs.

vojtechjelinek commented on scripts/initial_release_prep.py:
```diff
@@ -0,0 +1,223 @@
+            'Email Address',
+            'What is your name? ',
+            'Please give a clear description of why you want to '
+            'run this job on the test server')
```
Do we need to use these questions directly? Isn't there any possibility to change these headers to some more explaining variable like names (like `email_address` or `release_month`)? I imagine that this could be done in the sheet but the related form could still have full questions.

ankita240796 replied:
I am not sure about this. I am fine with changing the names in sheets but do we need to change the related form too then?

@seanlip @nithusha21 @DubeySandeep thoughts?

nithusha21 replied:
I think this is directly from the questions on the form. I am not sure if we can edit the title of the columns which are filled in automatically based on the form responses.

vojtechjelinek replied:
I think it is worth a try

ankita240796 replied:
I have added this is as an AI in the release process doc.

vojtechjelinek commented on scripts/initial_release_prep.py:
```diff
@@ -0,0 +1,223 @@
+            'Email Address',
+            'What is your name?',
+            'URL of a Google Doc with clear instructions on what the tester '
+            'needs to do:')
```
ditto

ankita240796 replied:
Replied above.

vojtechjelinek commented on scripts/update_changelog_and_credits.py:
```diff
@@ -459,22 +460,50 @@ def main():
+        '(Confirm from Sean in case of doubt).')
+    common.open_new_tab_in_browser_if_possible(
+        'https://docs.google.com/forms/d/'
+        '1yH6ZO2UiD_VspgKJR40byRSjUP1AaBF9ARSe814p8K0/edit#responses')
```
Move this into a constant?

ankita240796 replied:
Done.

vojtechjelinek reviewed (changes requested):
Thanks! Few questions/comments. I'm also thinking if we should move all the release scripts into subfolder in `scripts`.

kevinlee12 reviewed (commented):
just a comment!

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into doc-automate

ankita240796 committed:
Fix unicode issue

ankita240796 committed:
Address review comments

ankita240796 committed:
Add automation for schema changes detection, audio language change detection

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into doc-automate

ankita240796 commented:
Hi @vojtechjelinek @kevinlee12 I have addressed the review comments, PTAL. Thanks!

ankita240796 committed:
Strict check for extracting PR numbers

ankita240796 committed:
Fix tests

kevinlee12 commented on .github/CODEOWNERS:
```diff
@@ -101,7 +101,50 @@
+/scripts/pre_commit_linter.py @kevinlee12
+/scripts/linter_utils_test.py @kevinlee12
+/scripts/clean.py @kevinlee12
+/scripts/deploy.py @kevinlee12
```
@ankita240796 Would you mind also jointly, owning this? I'm not too familiar with the code that occurs here.

ankita240796 replied:
Done.

kevinlee12 commented on .github/CODEOWNERS:
```diff
@@ -101,7 +101,50 @@
+/scripts/clean.py @kevinlee12
+/scripts/deploy.py @kevinlee12
+/scripts/pre_commit_hook.py @kevinlee12
+/scripts/create_topological_sort_of_all_services.py @kevinlee12
```
Please also add @bansalnitish for this file.

ankita240796 replied:
Done.

kevinlee12 commented on .github/CODEOWNERS:
```diff
@@ -101,7 +101,50 @@
 /core/templates/dev/head/css/.stylelintrc @kevinlee12
 /tox.ini @kevinlee12
-/scripts/ @kevinlee12
+/scripts/install_third_party_libs.py @kevinlee12
```
Please also alphabetize this section.

ankita240796 replied:
Done.

kevinlee12 reviewed (commented):
Couple more code owners comments, ptal!

ankita240796 committed:
Fix codeowners

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into doc-automate

nithusha21 commented on scripts/initial_release_prep.py:
```diff
@@ -0,0 +1,312 @@
+        '2. Navigate to the admin panel and then the jobs tab\n'
+        '3. Run the above jobs\n'
+        '4. In case of failure/success, please send the output logs for '
+        'the job to me and the job author:\n\n'
```
```suggestion
        'the job to me and %s:\n\n'
```
Add the authors name as I think we have that data directly right? You can also maybe add the author's email id.

ankita240796 replied:
The author name is mentioned in the list of the jobs ([L109](https://github.com/oppia/oppia/pull/8005/files/60b8dad47d0d646f925dce41ea75a405c9d925c3#diff-5ec19cb5334ecdb7f1dada588baf5b5eR109), [L116](https://github.com/oppia/oppia/pull/8005/files/60b8dad47d0d646f925dce41ea75a405c9d925c3#diff-5ec19cb5334ecdb7f1dada588baf5b5eR116)) already. So, should we repeat it here?

nithusha21 replied:
Makes it easier for Sean😅

ankita240796 replied:
That's a good point 😛 

Updated!

nithusha21 commented on scripts/initial_release_prep.py:
```diff
@@ -0,0 +1,312 @@
+    # First three letters of the release month. Eg: Nov, Dec etc.
+    current_release_month = datetime.datetime.utcnow().strftime('%h')
+    for job_detail in job_details:
+        if current_release_month in job_detail[month_header]:
```
Maybe we'll need to make the month field a dropdown in the form (as not the author is allowed to type in anything I believe). Also, maybe we should do a tolower() before checking?

ankita240796 replied:
> Maybe we'll need to make the month field a dropdown in the form (as not the author is allowed to type in anything I believe)

Yes, we should do that.

> Also, maybe we should do a tolower() before checking?

Done.

nithusha21 replied:
Can you add an AI somewhere (in the release process doc maybe) to fix the form?

ankita240796 replied:
Done.

ankita240796 replied:
Done.

nithusha21 commented on scripts/initial_release_prep.py:
```diff
@@ -0,0 +1,312 @@
+        from constants import constants
+        supported_audio_languages_for_previous_release = constants[
+            'SUPPORTED_AUDIO_LANGUAGES']
+    finally:
```
What's the try block doing here? And why do we expect an exception?

ankita240796 replied:
The command [here](https://github.com/oppia/oppia/pull/8005/files/60b8dad47d0d646f925dce41ea75a405c9d925c3#diff-5ec19cb5334ecdb7f1dada588baf5b5eR182) converts the constants file to a version from previous release and stages it as well (`git add`). So, the try finally is added to be safe in case of any random exception i.e. the changes to constants file are reverted even in case the function faces an exception.

nithusha21 replied:
Ah I see.. Have you run into an exception here while testing it?

ankita240796 replied:
No, I did not get any specific exception while testing.

nithusha21 commented on scripts/initial_release_prep.py:
```diff
@@ -0,0 +1,312 @@
+            'SUPPORTED_AUDIO_LANGUAGES']
+    finally:
+        common.run_cmd(['git', 'reset', 'assets/constants.ts'])
+        common.run_cmd(['git', 'checkout', '--', 'assets/constants.ts'])
```
Won't both these commands do the same thing?

ankita240796 replied:
The constants file is staged due to [this command](https://github.com/oppia/oppia/pull/8005/files/60b8dad47d0d646f925dce41ea75a405c9d925c3#diff-5ec19cb5334ecdb7f1dada588baf5b5eR182). The reset command unstages it but the changes are still present. The checkout command removes the changes as well.

nithusha21 commented on scripts/release_info.py:
```diff
@@ -139,7 +143,7 @@ def extract_pr_numbers(logs):
     """
     pr_numbers = PR_NUMBER_REGEX.findall(
-        ' '.join([log.message for log in logs]))
+        ' '.join([log.message.split('\n')[0] for log in logs]))
```
Can we just do one more thing to the regex? Can we make sure that the (#XXXX) is at the end of the line? (Add a $ at the end of the regex). This would make the regex not match (#xxxx) patterns in the title of the PR.

ankita240796 replied:
Done.

nithusha21 commented on scripts/update_changelog_and_credits_test.py:
```diff
@@ -99,6 +100,12 @@ def mock_run_cmd(unused_cmd):
+        def mock_main(unused_personal_access_token):
+            pass
+        # pylint: disable=unused-argument
+        def mock_getpass(prompt):
```
Can't you do unused_prompt and avoid the pragma?

ankita240796 replied:
`prompt` is passed as a named argument in the `getpass` function, so we cannot change it to `unused_prompt`.

nithusha21 reviewed (commented):
Overall LGTM, left some comments!

kevinlee12 reviewed (approved):
lgtm!

ankita240796 committed:
Make regex strict, fix month check

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into doc-automate

ankita240796 committed:
Add author names and mail ids

vojtechjelinek commented on scripts/deploy.py:
```diff
@@ -319,10 +317,6 @@ def check_breakage(app_name, current_release_version):
     test_server_error_logs_url = (
         'https://console.cloud.google.com/logs/viewer?'
         'project=%s&key1=default&minLogLevel=500') % app_name
```
Wouldn't it make sense to move this to constants too?

ankita240796 replied:
Not sure, it is kind of a variable since it depends on app name and is specific to the function as well. What do you think?

vojtechjelinek commented on release_constants.py:
```diff
@@ -41,3 +41,27 @@
+    'https://docs.google.com/document/d/'
+    '1pmcDNfM2KtmkZeYipuInC48RE5JfkSJWQYdIQAkD0hQ/edit#')
+
+EXISTING_JOB_SPREADSHEETS_URL = (
```
```suggestion
EXISTING_JOBS_SPREADSHEETS_URL = (
```

ankita240796 replied:
Updated.

vojtechjelinek commented on release_constants.py:
```diff
@@ -41,3 +41,27 @@
+EXISTING_JOB_SPREADSHEETS_URL = (
+    'https://docs.google.com/spreadsheets/d/'
+    '1cSoVEwFyT-Q6d7yonbB0N-ElYGCFz5feMWXZXsv16_Y/edit#gid=1262496785')
+NEW_JOBS_SPREADSHEET_URL = (
```
I would change `NEW_JOBS` to something else, like `ONE_TIME_JOBS` because now it seems to me that these jobs will then be moved to the `EXISTING_JOBS`, it would maybe even make sense to call `EXISTING_JOBS` somethin like `REPEATABLE_JOBS` or `REPETITIVE_JOBS`.

ankita240796 replied:
Done.

vojtechjelinek commented on scripts/deploy.py:
```diff
@@ -374,10 +370,14 @@ def check_travis_and_circleci_tests(current_branch_name):
 
-    if requests.get(travis_url).status_code != 200:
+    try:
+        python_utils.url_open(travis_url)
+    except Exception:
         travis_url = 'https://travis-ci.org/oppia/oppia/branches'
 
-    if requests.get(circleci_url).status_code != 200:
+    try:
+        python_utils.url_open(circleci_url)
+    except Exception:
```
Should we catch more specific `Exception` here?

ankita240796 replied:
Not sure - since the [function](https://github.com/oppia/oppia/blob/develop/python_utils.py#L295) itself uses `urllib2` or `urllib.request` based on python version, so it will be harder to catch exceptions. Also, I think it is ok to redirect the user to Oppia travis/circleci urls in case of any exception since that will definitely exist.

nithusha21 reviewed (approved):
LGTM, just left a couple of follow up comments.

vojtechjelinek commented on scripts/initial_release_prep.py:
```diff
@@ -0,0 +1,317 @@
+    if check_changes_in_supported_audio_languages(
+            remote_alias, previous_release_version):
+        extra_jobs_to_run.append(
+            'ExplorationOpportunitySummaryModelRegenerationJob')
```
Could you add a comment why is this run?

ankita240796 replied:
Added.

vojtechjelinek commented on scripts/initial_release_prep_test.py:
```diff
@@ -0,0 +1,318 @@
+            'job_name': 'Job2',
+            'author_email': 'alice@email.com',
+            'author_name': 'Alice',
+            'instruction_doc': 'Instructions for job 2'
```
Maybe use some dummy URLs in the `instruction_doc` field (it shouldn't break anything but I think that it makes sense to get as close to the real data as possible.

ankita240796 replied:
Done.

vojtechjelinek commented on scripts/restore_backup.py:
```diff
@@ -95,20 +118,23 @@ def main(args=None):
 
-    if options.check_status:
-        check_backup_restoration_status()
+    if options.check_status or options.cancel_operation:
+        if options.check_status:
+            check_backup_restoration_status()
+        if options.cancel_operation:
+            cancel_operation()
```
Why not:
```python
if options.check_status:
    check_backup_restoration_status()
elif options.cancel_operation:
    cancel_operation()
```
?

ankita240796 replied:
Just for a case when user uses both flags to check status as well as cancel operation.

vojtechjelinek reviewed (changes requested):
Thanks! Took a full pass. What about moving the release related scripts into separate subfolder?

ankita240796 committed:
Minor fixes based on current release

ankita240796 committed:
Address review comments

ankita240796 committed:
Separate release related scripts

ankita240796 commented:
Hi @vojtechjelinek, I have addressed all the review comments, PTAL. Thanks!

> What about moving the release related scripts into separate subfolder?

Done.

ankita240796 committed:
Fix branch cut

seanlip commented on scripts/release_scripts/cut_release_or_hotfix_branch.py:
```diff
@@ -189,48 +175,71 @@ def _verify_hotfix_number_is_one_ahead_of_previous_hotfix_number(
+        new_branch_type = release_constants.RELEASE_BRANCH_TYPE
+    else:
+        new_branch_name = '%s-%s-%s-%s' % (
+            release_constants.RELEASE_BRANCH_TYPE, target_version,
```
It seems a bit weird to use constants of the form RELEASE_BRANCH_TYPE and HOTFIX_BRANCH_TYPE. These sound like two separate branch types but they are being used in the same branch name.

I suggest that you do the "cut" somewhere else, based on what is needed by the different functions. For example you might instead define private methods called _get_hotfix_branch_prefix() or _get_release_branch_name() or _get_hotfix_branch_name() (these are just examples, do whatever you need to in order to satisfy the needs of the functions that manipulate branch names).

ankita240796 replied:
Done.

seanlip commented on scripts/release_scripts/cut_release_or_hotfix_branch_test.py:
```diff
@@ -0,0 +1,395 @@
+import python_utils
+from scripts import common
+
+from . import cut_release_or_hotfix_branch
```
Possible to do `from scripts.release_scripts` (and similarly elsewhere)? Prefer to be fully explicit with imports where possible.

ankita240796 replied:
Done.

seanlip commented on scripts/release_scripts/cut_release_or_hotfix_branch_test.py:
```diff
@@ -0,0 +1,395 @@
+            'ERROR: Failed to fetch latest release info from GitHub.'):
+            (
+                cut_release_or_hotfix_branch
+                .verify_target_version_is_consistent_with_latest_released_version( # pylint: disable=line-too-long
```
Let's try and shorten this name. 

verify_target_version_compatible_with_latest_release perhaps -- does that help?

ankita240796 replied:
Yes, updated.

seanlip commented on scripts/release_scripts/initial_release_prep.py:
```diff
@@ -0,0 +1,321 @@
+                'author_email': job_detail[author_email_header],
+                'author_name': job_detail[author_name_header],
+                'instruction_doc': job_detail[instruction_doc_header]
+                })
```
Deindent by 4 to match opening indent.

/cc @oppia/dev-workflow-team this should be a lint check.

ankita240796 replied:
Done.

seanlip commented on scripts/release_scripts/initial_release_prep.py:
```diff
@@ -0,0 +1,321 @@
+    return list(set(jobs_to_run))
+
+
+def check_changes_in_supported_audio_languages(
```
did_supported_audio_languages_change() or have_supported_audio_languages_changed() 

(to imply that a boolean is being returned)

ankita240796 replied:
Done.

seanlip commented on scripts/release_scripts/initial_release_prep.py:
```diff
@@ -0,0 +1,321 @@
+        release_constants.RELEASE_NOTES_URL)
+    common.ask_user_to_confirm(
+        'Please check if anything extra is required for the release '
+        'and keep a track of this and do it at the appropriate point.')
```
... for the release. If so, keep track of this and ...

ankita240796 replied:
Done.

seanlip commented on scripts/release_scripts/release_info.py:
```diff
@@ -108,7 +107,11 @@ def gather_logs(start, stop='HEAD'):
     get_logs_cmd = GIT_CMD_GET_LOGS_FORMAT_STRING.format(
         GROUP_SEP, start, stop)
-    out = common.run_cmd(get_logs_cmd.split(' ')).split('\x00')
+    # The unicode conversion is required because there can be non-ascii
```
There should be some tests for this (i.e. that involve unicode chars).

ankita240796 replied:
Added.

seanlip commented on scripts/release_scripts/release_info.py:
```diff
@@ -28,8 +28,7 @@
 
```
Can we rename this script to generate_release_info.py? (Similar to other scripts which start with a verb.)

ankita240796 replied:
Done.

seanlip commented on scripts/release_scripts/restore_backup.py:
```diff
@@ -86,6 +97,17 @@ def check_backup_restoration_status():
+    python_utils.PRINT('List of operations in progress:\n')
+    check_backup_restoration_status()
+    python_utils.PRINT(
+        'Enter the name of the operation to cancel.')
```
This is a bit odd. What exactly should I enter here? Some more guidance might be appropriate.

Also, we should show a warning if cancellation happens because that means the server might be in a corrupt state? Or does no data actually get transferred until the very end?

ankita240796 replied:
> This is a bit odd. What exactly should I enter here? Some more guidance might be appropriate.

I will need some help on this. We need to check this when doing backup restoration.

From the example [here](https://stackoverflow.com/a/53630367), I think the list operation should print out ids (names) of operation which can be entered by the user for cancellation. But we need to confirm this.

> Also, we should show a warning if cancellation happens because that means the server might be in a corrupt state? Or does no data actually get transferred until the very end?

I think the server is in a corrupt state if we cancel ([Ref](https://cloud.google.com/datastore/docs/export-import-entities#cancel_an_operation)). Should I add a warning message then?

seanlip replied:
Yep, please add the warning and include the link you just sent in the warning too.

For the first part -- I'm not sure exactly what help you need, but feel free to ping me / send me instructions and I'll do it!

ankita240796 replied:
> Yep, please add the warning and include the link you just sent in the warning too.

Added.

> For the first part -- I'm not sure exactly what help you need, but feel free to ping me / send me instructions and I'll do it!

We will need to run a backup restoration and check what is the output from `check_status` function. Is this possible now?

seanlip commented on scripts/release_scripts/update_changelog_and_credits.py:
```diff
@@ -33,8 +33,9 @@
+from scripts import common
 
-from . import common
+from . import release_info
```
from scripts import release_info

ankita240796 replied:
Done.

seanlip commented on scripts/release_scripts/update_changelog_and_credits.py:
```diff
@@ -435,7 +438,11 @@ def create_branch(repo_fork, target_branch, github_username):
         'Pushed changes to Github. '
-        'Please create a pull request from the %s branch' % target_branch)
+        'Please create a pull request from the %s branch\n\n'
+        'Note: PR title should be exactly: '
```
I think you might be able to just provide a URL with the title in it? Here's an example:

https://github.com/oppia/oppia/compare/develop...feature-branch?quick_pull=1&title=this-is-awesome

(change the title and the source/target branch names appropriately).

ankita240796 replied:
Done.

seanlip commented on scripts/release_scripts/update_changelog_and_credits.py:
```diff
@@ -459,22 +460,49 @@ def main():
+        'Check emails and names for authors and contributors and '
+        'verify that the emails are '
+        'correct through welcome emails sent from welcome@oppia.org '
+        '(Confirm from Sean in case of doubt).')
```
`(confirm with ...)`

ankita240796 replied:
Done.

seanlip commented on scripts/release_scripts/update_changelog_and_credits.py:
```diff
@@ -459,22 +460,49 @@ def main():
+    common.open_new_tab_in_browser_if_possible(
+        release_constants.CREDITS_FORM_URL)
+    common.ask_user_to_confirm(
+        'Check the credits form and add any non technical contributors '
```
non technical --> additional

ankita240796 replied:
Done.

seanlip commented on scripts/release_scripts/update_changelog_and_credits.py:
```diff
@@ -459,22 +460,49 @@ def main():
+        'Check the credits form and add any non technical contributors '
+        'to the contributor list in release summary file.')
+    common.ask_user_to_confirm(
+        'Categorize the terms in uncategorized section and arrange '
```
"terms" sounds like it refers to "terms and conditions" pages.

Let's make it "Categorize the PR titles in the Uncategorized section of the changelog, and arrange the changelog to have user-facing categories on top."

ankita240796 replied:
Done.

seanlip commented on scripts/release_scripts/update_changelog_and_credits.py:
```diff
@@ -459,22 +460,49 @@ def main():
+        'Categorize the terms in uncategorized section and arrange '
+        'the changelog to have user facing categories on top.')
+    common.ask_user_to_confirm(
+        'Verify each item is in correct section and remove '
```
in correct --> in the correct

ankita240796 replied:
Done.

seanlip commented on scripts/release_scripts/update_configs.py:
```diff
@@ -168,19 +168,21 @@ def add_mailgun_api_key():
 
 
-def main():
+def main(personal_access_token):
```
Add args to docstring.

ankita240796 replied:
Done.

seanlip commented on scripts/release_scripts/wrap_up_release_test.py:
```diff
@@ -25,9 +25,9 @@
 
-from . import common
-from . import wrap_release
+from . import wrap_up_release
```
from scripts.release_scripts

ankita240796 replied:
Done.

seanlip reviewed (changes requested):
Thanks @ankita240796! Took a full pass.

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into doc-automate

ankita240796 committed:
Address review comments

ankita240796 commented:
Hi @seanlip, I have addressed all the review comments, PTAL. Thanks!

seanlip commented on scripts/release_scripts/cut_release_or_hotfix_branch.py:
```diff
@@ -194,6 +194,36 @@ def verify_hotfix_number_is_one_ahead_of_previous_hotfix_number(
+        tuple(str, str). The type and name of release branch.
+    """
+    return (
+        release_constants.RELEASE_BRANCH_TYPE, '%s-%s' % (
```
This should be:

    return (release_constants.BRANCH_TYPE_RELEASE, 'release-%s' % target_version)

(When naming constants, start with the type. Also don't confuse the concepts of branch type and branch name.)

ankita240796 replied:
Done.

seanlip commented on scripts/release_scripts/cut_release_or_hotfix_branch.py:
```diff
@@ -194,6 +194,36 @@ def verify_hotfix_number_is_one_ahead_of_previous_hotfix_number(
+        tuple(str, str). The type and name of hotfix branch.
+    """
+    return (
+        release_constants.HOTFIX_BRANCH_TYPE, '%s-%s-%s-%s' % (
```
Similarly:

    return (release_constants.BRANCH_TYPE_HOTFIX, 'release-%s-hotfix-%s' % ...)

ankita240796 replied:
Done.

seanlip commented on scripts/release_scripts/update_changelog_and_credits.py:
```diff
@@ -435,7 +431,8 @@ def create_branch(
     common.open_new_tab_in_browser_if_possible(
         'https://github.com/oppia/oppia/compare/develop...%s:%s?'
-        'expand=1' % (github_username, target_branch))
+        'expand=1&title=Update authors and changelog for v%s' % (
```
Just to check, did you manually test this URL? (You can try it with an existing branch.)

Wanted to make sure the encoding is correct.

ankita240796 replied:
Yes, tested. It works fine.

seanlip reviewed (changes requested):
Thanks! Just a few more notes.

vojtechjelinek reviewed (approved):
LGTM!

ankita240796 committed:
Fix issues with release info generation for hotfixes

ankita240796 committed:
Address review comments

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into doc-automate

DubeySandeep commented on scripts/release_scripts/cut_release_or_hotfix_branch.py:
```diff
@@ -46,8 +46,7 @@
 import release_constants
-
-from . import common
+from scripts import common
```
are we expecting this to be above `import python_utils`?

ankita240796 replied:
No, in sorted order it will come after `import python_utils` (`P` -> `python_utils` then `S` -> `scripts`)

DubeySandeep commented on scripts/release_scripts/cut_release_or_hotfix_branch.py:
```diff
@@ -113,21 +95,24 @@ def _verify_target_branch_does_not_already_exist(remote_alias):
+        target_version):
     """Checks that the target version is consistent with the latest released
     version on GitHub.
+    Args:
```
Add a newline above^

ankita240796 replied:
Added.

DubeySandeep commented on scripts/release_scripts/cut_release_or_hotfix_branch.py:
```diff
@@ -189,48 +175,94 @@ def _verify_hotfix_number_is_one_ahead_of_previous_hotfix_number(
+        tuple(str, str). The type and name of release branch.
+    """
+    return (
+        release_constants.RELEASE_BRANCH_TYPE, 'release-%s' % target_version)
```
Are we expecting the hardcoded `release-%s` to chang on changing the value of `RELEASE_BRANCH_TYPE`? If yes, then use the constant.

ankita240796 replied:
This is following @seanlip's [comment](https://github.com/oppia/oppia/pull/8005#discussion_r351038440) to avoid confusion between type and name. I actually missed updating this at a few places. Fixed now.

DubeySandeep commented on scripts/release_scripts/cut_release_or_hotfix_branch.py:
```diff
@@ -189,48 +175,94 @@ def _verify_hotfix_number_is_one_ahead_of_previous_hotfix_number(
+        tuple(str, str). The type and name of hotfix branch.
+    """
+    return (
+        release_constants.HOTFIX_BRANCH_TYPE, 'release-%s-hotfix-%s' % (
```
Ditto here I would suggest changing this similar to line 180.

ankita240796 replied:
Replied above.

DubeySandeep commented on scripts/release_scripts/cut_release_or_hotfix_branch.py:
```diff
@@ -189,48 +175,94 @@ def _verify_hotfix_number_is_one_ahead_of_previous_hotfix_number(
-            branch_to_check = '%s-%s' % (
-                release_constants.RELEASE_BRANCH_TYPE, TARGET_VERSION)
+        elif hotfix_number == 1:
+            branch_to_check = 'release-%s' % target_version
```
Should we use `release_constants.RELEASE_BRANCH_TYPE`?

ankita240796 replied:
Replied above.

DubeySandeep commented on scripts/release_scripts/cut_release_or_hotfix_branch.py:
```diff
@@ -189,48 +175,94 @@ def _verify_hotfix_number_is_one_ahead_of_previous_hotfix_number(
-            branch_to_check = '%s-%s-%s-%s' % (
-                release_constants.RELEASE_BRANCH_TYPE, TARGET_VERSION,
-                release_constants.HOTFIX_BRANCH_TYPE, HOTFIX_NUMBER - 1)
+            branch_to_check = 'release-%s-hotfix-%s' % (
```
Ditto*

ankita240796 replied:
Replied above.

DubeySandeep commented on scripts/release_scripts/cut_release_or_hotfix_branch.py:
```diff
@@ -244,33 +276,36 @@ def _execute_branch_cut():
+        verify_hotfix_number_is_one_ahead_of_previous_hotfix_number(
+            remote_alias, target_version, hotfix_number)
+        if hotfix_number == 1:
+            branch_to_cut_from = 'release-%s' % target_version
```
I'm not sure why we are hardcoding this value somewhere and somewhere we are using the constant, am I missing something?

ankita240796 replied:
Replied above.

DubeySandeep commented on scripts/release_scripts/deploy.py:
```diff
@@ -49,13 +49,10 @@
+from scripts import common
+from scripts import install_third_party_libs
+from scripts.release_scripts import gcloud_adapter
+from scripts.release_scripts import update_configs
```
Move these imports above line 50. 

(Also, will it be possible to use pragma just for the argparse imports, can you please check?)

ankita240796 replied:
No, it gives error if we use the pragma just for argparse.

seanlip commented on scripts/release_scripts/generate_release_info.py:
```diff
@@ -67,7 +67,10 @@ def get_current_version_tag(repo):
     """
-    return repo.get_tags()[0]
+    if 'hotfix' in common.get_current_branch_name():
+        return repo.get_tags()[1]
```
It might be possible to have more than one hotfix for a given release. Will this still work in that case? (Please add tests for this case too.)

ankita240796 replied:
Yes, it will since in case of multiple hotfixes as well, the first tag will be of current release and the second of previous release.

(For example, for release-2.8.7, the tag will be v2.8.7, now even if we have two hotfixes, the first tag is v2.8.7 and second v2.8.6, so this function will return v2.8.6 irrespective of the number of hotfix)

Added a test as well.

seanlip replied:
As discussed, please add a detailed comment explaining conceptually what's going on here and why [1] and [0] are the correct indexes to look at respectively.

ankita240796 replied:
Added.

seanlip commented on scripts/common.py:
```diff
@@ -144,6 +148,21 @@ def get_current_branch_name():
+    Returns:
+        str. The version of release.
+    """
+    release_version = release_branch_name[len(RELEASE_BRANCH_NAME_PREFIX):]
```
Optional: If possible it would be cleaner to do things more semantically rather than using manual operations with indexes. For example I wonder if you can compare the branch name to a regex instead, and use that to extract the relevant "pieces" from it.

(This is optional, don't spend too much time on it. But it would improve readability.)

ankita240796 replied:
Done.

seanlip reviewed (commented):
Just a couple notes inline, otherwise LGTM!

Re the backup -- I would suggest, for now, clarifying the instructions according to the SO post (i.e. explain exactly what thing from the list should be typed in. We can do the live check when we meet in a couple of days, but that post seems correct to me from what I remember.

seanlip commented:
@DubeySandeep PTAL at the parts of this PR that correspond to new language addition. Thanks!

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into doc-automate

ankita240796 committed:
Address review comments

ankita240796 commented:
>Re the backup -- I would suggest, for now, clarifying the instructions according to the SO post (i.e. explain exactly what thing from the list should be typed in. We can do the live check when we meet in a couple of days, but that post seems correct to me from what I remember.

Done.

Hi @seanlip, I have addressed all the review comments, PTAL. Thanks!

DubeySandeep commented on scripts/release_scripts/deploy.py:
```diff
@@ -398,6 +397,22 @@ def check_travis_and_circleci_tests(current_branch_name):
+    common.ask_user_to_confirm(
+        'Please create a dedicated section for this release in the '
+        'release tracking document created by the QA Lead.\n'
+        'The three tabs in your browser point to: '
```
Do we want to show this message even if `open_new_tab_in_browser_if_possible` wasn't able to open the browser?

ankita240796 replied:
Yes, I think it's ok since even if tabs cannot be opened automatically, the function `open_new_tab_in_browser_if_possible` asks user to open tabs before proceeding further.

seanlip reviewed (approved):
Thanks, @ankita240796! Just one remaining comment but otherwise LGTM from my end.

seanlip commented:
(Feel free to fix, and then merge.)

DubeySandeep commented on scripts/release_scripts/initial_release_prep.py:
```diff
@@ -0,0 +1,321 @@
+        # contributors to translate and voiceover an exploration
+        # whenever a new audio language is added.
+        # Refer: https://github.com/oppia/oppia/issues/8027.
+        extra_jobs_to_run.append(
```
We want to run this job only if a new language is added or we delete a language, but `did_supported_audio_languages_change` will trigger this on any changes in the languages_ list. Should we show the changes and let the user decided whether we need to run the job? 

(This will may create confusion and a complicated workflow, I'm fine with the current workflow but wanted to know whether it's fine to run the job sometime (Which I don't think will happen frequently) when it's not required in cost of having a simple and clear workflow. @seanlip any suggestion?)

DubeySandeep replied:
<Ignore this comment, I've suggested a better way to handel this in the comment below.>

DubeySandeep commented on scripts/release_scripts/initial_release_prep.py:
```diff
@@ -0,0 +1,321 @@
+        common.run_cmd(['git', 'checkout', '--', 'assets/constants.ts'])
+
+    return (
+        sorted(supported_audio_languages_for_current_release) != sorted(
```
Can we just check the `id` of each item in the list instead of equating the whole dict?

(We need to run the job only if a new id is added or deleted.)

ankita240796 replied:
Done.

DubeySandeep reviewed (approved):
@ankita240796, Thanks for adding the check for the auppported_audio languages changes. This looks great! :)

Also, just to understand the new workflow I also took a quick look into a few files and left a few questions/comments, please take a look!

DubeySandeep reviewed (changes requested):
Marking this as request changes so that the audio_language changes related comments are resolved before someone merges this PR.

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into doc-automate

ankita240796 committed:
Update comment

ankita240796 committed:
Add newline

ankita240796 committed:
Fix names

ankita240796 committed:
Fix language check

ankita240796 committed:
Fix version issues

ankita240796 commented:
Hi @seanlip @DubeySandeep PTAL, Thanks!

seanlip commented on scripts/release_scripts/generate_release_info.py:
```diff
@@ -67,6 +67,12 @@ def get_current_version_tag(repo):
+    # release for which the hotfix branch is. So, if we require generation
+    # of release summary in case of hotfix, we need to get the second version
+    # tag. For example, if branch is release-1.2.3-hotfix-1, the first tag
+    # on github will 1.2.3 but the correct tag of previous release will be
```
will --> will be

ankita240796 replied:
Done.

seanlip commented on scripts/release_scripts/generate_release_info.py:
```diff
@@ -67,6 +67,12 @@ def get_current_version_tag(repo):
+    # 1.2.2 which is required for release summary generation.
     if 'hotfix' in common.get_current_branch_name():
         return repo.get_tags()[1]
     else:
```
Explain this case too. Why [0]?

ankita240796 replied:
Done.

seanlip reviewed (commented):
Thanks @ankita240796! I'll defer to @DubeySandeep to review the language stuff. I just have two small comments.

ankita240796 committed:
Fix comments

seanlip commented:
LGTM. @DubeySandeep PTAL!

DubeySandeep reviewed (approved):
LGTM, for the supported audio language change feature! :)
